agent/unit/graph_test.go

```diff
@@ -0,0 +1,653 @@
+		assertDOTGraph(t, graph, "SelfReference")
+	})
+
+	t.Run("Cycle", func(t *testing.T) {
+		t.Parallel()
+
+		graph := &unit.Graph[Status, *Unit]{}
+		unit1 := &Unit{Name: "unit1", Status: StatusPending}
+		unit2 := &Unit{Name: "unit2", Status: StatusPending}
+		err := graph.AddEdge(unit1, unit2, StatusCompleted)
+		require.NoError(t, err)
+		err = graph.AddEdge(unit2, unit1, StatusStarted)
+		require.Error(t, err)
+		require.ErrorContains(t, err, fmt.Sprintf("adding edge (%v -> %v) would create a cycle", unit2, unit1))
+
+		// Assert on the DOT representation using golden file (should contain only the first edge)
+		assertDOTGraph(t, graph, "Cycle")
+	})
+
+	t.Run("MultipleDependenciesSameStatus", func(t *testing.T) {
+		t.Parallel()
+
+		graph := &unit.Graph[Status, *Unit]{}
+		unit1 := &Unit{Name: "unit1", Status: StatusPending}
+		unit2 := &Unit{Name: "unit2", Status: StatusPending}
+		unit3 := &Unit{Name: "unit3", Status: StatusPending}
+		unit4 := &Unit{Name: "unit4", Status: StatusPending}
+
+		// Unit1 depends on completion of both unit2 and unit3 (same status type)
+		err := graph.AddEdge(unit1, unit2, StatusCompleted)
+		require.NoError(t, err)
+		err = graph.AddEdge(unit1, unit3, StatusCompleted)
+		require.NoError(t, err)
+
+		// Unit1 also depends on starting of unit4 (different status type)
+		err = graph.AddEdge(unit1, unit4, StatusStarted)
+		require.NoError(t, err)
+
+		// Check that unit1 has 3 forward dependencies
+		forwardEdges := graph.GetForwardAdjacentVertices(unit1)
+		require.Len(t, forwardEdges, 3)
+
+		// Verify all expected dependencies exist
+		expectedDependencies := []TestEdge{
+			{From: unit1, To: unit2, Edge: StatusCompleted},
+			{From: unit1, To: unit3, Edge: StatusCompleted},
+			{From: unit1, To: unit4, Edge: StatusStarted},
+		}
+
+		for _, expected := range expectedDependencies {
+			require.Contains(t, forwardEdges, expected)
+		}
+
+		// Check reverse dependencies
+		unit2ReverseEdges := graph.GetReverseAdjacentVertices(unit2)
+		require.Len(t, unit2ReverseEdges, 1)
+		require.Contains(t, unit2ReverseEdges, TestEdge{
+			From: unit1, To: unit2, Edge: StatusCompleted,
+		})
+
+		unit3ReverseEdges := graph.GetReverseAdjacentVertices(unit3)
+		require.Len(t, unit3ReverseEdges, 1)
+		require.Contains(t, unit3ReverseEdges, TestEdge{
+			From: unit1, To: unit3, Edge: StatusCompleted,
+		})
+
+		unit4ReverseEdges := graph.GetReverseAdjacentVertices(unit4)
+		require.Len(t, unit4ReverseEdges, 1)
+		require.Contains(t, unit4ReverseEdges, TestEdge{
+			From: unit1, To: unit4, Edge: StatusStarted,
+		})
+
+		// Assert on the DOT representation using golden file
+		assertDOTGraph(t, graph, "MultipleDependenciesSameStatus")
+	})
+}
+
+func TestGraphThreadSafety(t *testing.T) {
+	t.Parallel()
+
+	t.Run("ConcurrentAddEdge", func(t *testing.T) {
+		t.Parallel()
+
+		graph := &unit.Graph[Status, *Unit]{}
+		var wg sync.WaitGroup
+		const numGoroutines = 100
```

Comment:
These are a bit extreme, I think perhaps 10 should suffice.

It's also a good idea to increase the chance of collisions e.g. via:

```go
sync := make(chan struct{})
for range 10; go func() {
	// prepare ...
	<-sync
	// work ...
}
close(sync)
```